// .ignored_openclaw/dist/plugin-sdk/web-CYvm_tfx.js
import { a as WA_WEB_AUTH_DIR, g as webAuthExists, s as logWebSelfId, u as pickWebChannel } from "./accounts-CR5voC9r.js";
import "./reply-YQvtZDnf.js";
import "./paths-DVWx7USN.js";
import "./github-copilot-token-Cg0YPPSu.js";
import "./plugins-lP-tdfUx.js";
import "./fetch-DdmiNDLY.js";
import "./subsystem-loAAXGlc.js";
import "./retry-4dLTmoU_.js";
import "./send-BbiKF8n5.js";
import "./config-XLaWsyXv.js";
import "./command-format-DI57or3f.js";
import "./model-selection-fS3jvgBm.js";
import "./agent-scope-DchfDHou.js";
import "./manifest-registry-Bp7G4UCL.js";
import "./dock-WIjq0YQy.js";
import "./redact-Cdy8PC2e.js";
import "./errors-BCGY_vDh.js";
import "./channel-activity-CrjbdYzh.js";
import "./image-ops-BkpfOGN7.js";
import "./ssrf-CnX37IfG.js";
import "./fetch-guard-KnovH0Do.js";
import "./local-roots-CDdEY71x.js";
import "./ir-l_BPhDkP.js";
import "./chunk-FbIfxd34.js";
import "./message-channel-BFQTPrdc.js";
import "./bindings-BejLzmNy.js";
import "./markdown-tables-B_6oysHH.js";
import "./render-3_VJ30aR.js";
import "./tables-CLyS-Ard.js";
import "./send-BSy5yHOV.js";
import "./tool-images-BcL1yi64.js";
import "./target-errors-DYCfK-1F.js";
import "./send-BLr9eFQ7.js";
import "./send-Bnjm2l2x.js";
import "./runner-B97FG2-t.js";
import "./reply-prefix-OXCn2hxF.js";
import "./tokens-BTtwtM4o.js";
import "./skill-commands-D8K8RudA.js";
import "./skills-DP2wNi0P.js";
import "./chrome-nGiTE_n8.js";
import "./accounts-lUR9W-qK.js";
import "./accounts-C33JUvW2.js";
import "./deliver-Cn6iEmbA.js";
import "./sessions-DOpDuDMA.js";
import "./paths-B-NY-HdV.js";
import "./diagnostic-CoFzGZeN.js";
import "./store-BI_y0OEJ.js";
import "./pi-embedded-helpers-BfA03LD8.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-CgCMU47d.js";
import "./thinking-BpFZfHN9.js";
import "./image-C7MqZNfB.js";
import "./pi-model-discovery-CNP1dAqt.js";
import "./api-key-rotation-DjaMpXLQ.js";
import "./query-expansion-DUDPIeZq.js";
import "./commands-registry-CHUzZBfZ.js";
import "./diagnostic-session-state-CpzJb54O.js";
import "./manager-BCcxT0_f.js";
import "./send-J_6o0ZUO.js";
import "./proxy-CLEew2L6.js";
import "./resolve-route-C6oUS6do.js";
import "./replies-DOWu3vWG.js";
import "./outbound-attachment-BzdFjC3g.js";
import { n as sendMessageWhatsApp } from "./outbound-CmjzU0im.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session-BWDwkvlE.js";
import { t as loginWeb } from "./login-BzX26MtI.js";

export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
