// .ignored_openclaw/dist/bundled/boot-md/handler.js
import { s as resolveAgentWorkspaceDir, t as listAgentIds } from "../../agent-scope-PyLC-ZZu.js";
import "../../paths-C9do7WCN.js";
import { dt as isGatewayStartupEvent, r as defaultRuntime, t as createSubsystemLogger } from "../../subsystem-9nuH7eAI.js";
import { l as resolveAgentIdFromSessionKey } from "../../session-key-YGVrIVO5.js";
import "../../workspace-CxArQZC-.js";
import "../../model-selection-C6wZ4s0z.js";
import "../../github-copilot-token-BkwQAVvU.js";
import "../../env-BCSW4P-L.js";
import "../../boolean-mcn6kL0s.js";
import { n as SILENT_REPLY_TOKEN } from "../../tokens-CGafcE1g.js";
import { a as createDefaultDeps, i as agentCommand } from "../../pi-embedded-BDhvoWGL.js";
import "../../plugins-D93Gl3vX.js";
import "../../accounts-C2p3gWX0.js";
import "../../bindings-XIxTfg76.js";
import "../../send-DNoo-__b.js";
import "../../send-CGHWrWIs.js";
import "../../deliver-DMxY-UiH.js";
import "../../diagnostic-Bfq8INpY.js";
import "../../diagnostic-session-state-C0Sxjfox.js";
import "../../accounts-DTMKYA1I.js";
import "../../send-WED6JSOM.js";
import "../../image-ops-D_zTmnMF.js";
import "../../pi-model-discovery-C-yOXpma.js";
import "../../message-channel-DSGEYImB.js";
import "../../pi-embedded-helpers-BWNHLmOk.js";
import "../../config-8KUmruTM.js";
import "../../manifest-registry-BoNM8D8y.js";
import "../../dock-BQZ3dOVS.js";
import "../../chrome-B_oaL_VN.js";
import "../../ssrf-BnuIFCPj.js";
import "../../frontmatter-rZSrGiqd.js";
import "../../skills-BmmA4kHX.js";
import "../../redact-Dhp4XjZk.js";
import "../../errors-oW2IM9f0.js";
import "../../store-D1p2WpL_.js";
import { V as resolveMainSessionKey, d as updateSessionStore, s as loadSessionStore, z as resolveAgentMainSessionKey } from "../../sessions-BJ6QL7GG.js";
import "../../accounts-CcUE2V_t.js";
import { l as resolveStorePath } from "../../paths-CLWDvYDE.js";
import "../../tool-images-BYjb6VbO.js";
import "../../thinking-CJoHneR6.js";
import "../../image-DIjIaXgj.js";
import "../../reply-prefix-D9MRO3Gt.js";
import "../../manager-CxmUPBMF.js";
import "../../gemini-auth-iOIz_-mf.js";
import "../../fetch-guard-Py2e9YaA.js";
import "../../query-expansion-Co0bWU3Q.js";
import "../../retry-Drrk8tTZ.js";
import "../../target-errors-COtYfWxr.js";
import "../../chunk-CP5JrSGh.js";
import "../../markdown-tables-Dc5vS6O9.js";
import "../../local-roots-BL_0qvcy.js";
import "../../ir-B9Vab0HN.js";
import "../../render-loap2gRq.js";
import "../../commands-registry-C_No2z8N.js";
import "../../skill-commands-D9vTBxFO.js";
import "../../runner-BS2WkEnz.js";
import "../../fetch-B1nZSYJF.js";
import "../../channel-activity-KmaKrhP4.js";
import "../../tables-Bf1rDqhy.js";
import "../../send-z9O0u_9n.js";
import "../../outbound-attachment-D7270zuT.js";
import "../../send-Jzg2ogVt.js";
import "../../resolve-route-nRTJSWCj.js";
import "../../proxy-Bee2aKQk.js";
import "../../replies-C9sCMljO.js";
import fs from "node:fs/promises";
import path from "node:path";
import crypto from "node:crypto";

//#region src/gateway/boot.ts
function generateBootSessionId() {
	return `boot-${(/* @__PURE__ */ new Date()).toISOString().replace(/[:.]/g, "-").replace("T", "_").replace("Z", "")}-${crypto.randomUUID().slice(0, 8)}`;
}
const log$1 = createSubsystemLogger("gateway/boot");
const BOOT_FILENAME = "BOOT.md";
function buildBootPrompt(content) {
	return [
		"You are running a boot check. Follow BOOT.md instructions exactly.",
		"",
		"BOOT.md:",
		content,
		"",
		"If BOOT.md asks you to send a message, use the message tool (action=send with channel + target).",
		"Use the `target` field (not `to`) for message tool destinations.",
		`After sending with the message tool, reply with ONLY: ${SILENT_REPLY_TOKEN}.`,
		`If nothing needs attention, reply with ONLY: ${SILENT_REPLY_TOKEN}.`
	].join("\n");
}
async function loadBootFile(workspaceDir) {
	const bootPath = path.join(workspaceDir, BOOT_FILENAME);
	try {
		const trimmed = (await fs.readFile(bootPath, "utf-8")).trim();
		if (!trimmed) return { status: "empty" };
		return {
			status: "ok",
			content: trimmed
		};
	} catch (err) {
		if (err.code === "ENOENT") return { status: "missing" };
		throw err;
	}
}
function snapshotMainSessionMapping(params) {
	const agentId = resolveAgentIdFromSessionKey(params.sessionKey);
	const storePath = resolveStorePath(params.cfg.session?.store, { agentId });
	try {
		const entry = loadSessionStore(storePath, { skipCache: true })[params.sessionKey];
		if (!entry) return {
			storePath,
			sessionKey: params.sessionKey,
			canRestore: true,
			hadEntry: false
		};
		return {
			storePath,
			sessionKey: params.sessionKey,
			canRestore: true,
			hadEntry: true,
			entry: structuredClone(entry)
		};
	} catch (err) {
		log$1.debug("boot: could not snapshot main session mapping", {
			sessionKey: params.sessionKey,
			error: String(err)
		});
		return {
			storePath,
			sessionKey: params.sessionKey,
			canRestore: false,
			hadEntry: false
		};
	}
}
async function restoreMainSessionMapping(snapshot) {
	if (!snapshot.canRestore) return;
	try {
		await updateSessionStore(snapshot.storePath, (store) => {
			if (snapshot.hadEntry && snapshot.entry) {
				store[snapshot.sessionKey] = snapshot.entry;
				return;
			}
			delete store[snapshot.sessionKey];
		}, { activeSessionKey: snapshot.sessionKey });
		return;
	} catch (err) {
		return err instanceof Error ? err.message : String(err);
	}
}
async function runBootOnce(params) {
	const bootRuntime = {
		log: () => {},
		error: (message) => log$1.error(String(message)),
		exit: defaultRuntime.exit
	};
	let result;
	try {
		result = await loadBootFile(params.workspaceDir);
	} catch (err) {
		const message = err instanceof Error ? err.message : String(err);
		log$1.error(`boot: failed to read ${BOOT_FILENAME}: ${message}`);
		return {
			status: "failed",
			reason: message
		};
	}
	if (result.status === "missing" || result.status === "empty") return {
		status: "skipped",
		reason: result.status
	};
	const sessionKey = params.agentId ? resolveAgentMainSessionKey({
		cfg: params.cfg,
		agentId: params.agentId
	}) : resolveMainSessionKey(params.cfg);
	const message = buildBootPrompt(result.content ?? "");
	const sessionId = generateBootSessionId();
	const mappingSnapshot = snapshotMainSessionMapping({
		cfg: params.cfg,
		sessionKey
	});
	let agentFailure;
	try {
		await agentCommand({
			message,
			sessionKey,
			sessionId,
			deliver: false
		}, bootRuntime, params.deps);
	} catch (err) {
		agentFailure = err instanceof Error ? err.message : String(err);
		log$1.error(`boot: agent run failed: ${agentFailure}`);
	}
	const mappingRestoreFailure = await restoreMainSessionMapping(mappingSnapshot);
	if (mappingRestoreFailure) log$1.error(`boot: failed to restore main session mapping: ${mappingRestoreFailure}`);
	if (!agentFailure && !mappingRestoreFailure) return { status: "ran" };
	return {
		status: "failed",
		reason: [agentFailure ? `agent run failed: ${agentFailure}` : void 0, mappingRestoreFailure ? `mapping restore failed: ${mappingRestoreFailure}` : void 0].filter((part) => Boolean(part)).join("; ")
	};
}

//#endregion
//#region src/hooks/bundled/boot-md/handler.ts
const log = createSubsystemLogger("hooks/boot-md");
const runBootChecklist = async (event) => {
	if (!isGatewayStartupEvent(event)) return;
	if (!event.context.cfg) return;
	const cfg = event.context.cfg;
	const deps = event.context.deps ?? createDefaultDeps();
	const agentIds = listAgentIds(cfg);
	for (const agentId of agentIds) {
		const workspaceDir = resolveAgentWorkspaceDir(cfg, agentId);
		const result = await runBootOnce({
			cfg,
			deps,
			workspaceDir,
			agentId
		});
		if (result.status === "failed") {
			log.warn("boot-md failed for agent startup run", {
				agentId,
				workspaceDir,
				reason: result.reason
			});
			continue;
		}
		if (result.status === "skipped") log.debug("boot-md skipped for agent startup run", {
			agentId,
			workspaceDir,
			reason: result.reason
		});
	}
};

//#endregion
export { runBootChecklist as default };
