// .ignored_openclaw/dist/plugin-sdk/audio-preflight-zyinzDIG.js
import "./accounts-CR5voC9r.js";
import "./paths-DVWx7USN.js";
import "./github-copilot-token-Cg0YPPSu.js";
import "./plugins-lP-tdfUx.js";
import { Z as logVerbose, et as shouldLogVerbose } from "./subsystem-loAAXGlc.js";
import "./config-XLaWsyXv.js";
import "./command-format-DI57or3f.js";
import "./model-selection-fS3jvgBm.js";
import "./agent-scope-DchfDHou.js";
import "./manifest-registry-Bp7G4UCL.js";
import "./dock-WIjq0YQy.js";
import "./redact-Cdy8PC2e.js";
import "./errors-BCGY_vDh.js";
import "./image-ops-BkpfOGN7.js";
import "./ssrf-CnX37IfG.js";
import "./fetch-guard-KnovH0Do.js";
import "./local-roots-CDdEY71x.js";
import "./message-channel-BFQTPrdc.js";
import "./bindings-BejLzmNy.js";
import "./tool-images-BcL1yi64.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-B97FG2-t.js";
import "./skills-DP2wNi0P.js";
import "./chrome-nGiTE_n8.js";
import "./accounts-lUR9W-qK.js";
import "./accounts-C33JUvW2.js";
import "./sessions-DOpDuDMA.js";
import "./paths-B-NY-HdV.js";
import "./store-BI_y0OEJ.js";
import "./pi-embedded-helpers-BfA03LD8.js";
import "./thinking-BpFZfHN9.js";
import "./image-C7MqZNfB.js";
import "./pi-model-discovery-CNP1dAqt.js";
import "./api-key-rotation-DjaMpXLQ.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
