// .ignored_openclaw/dist/audio-preflight-C8y3KBYy.js
import "./agent-scope-DKbLFdDh.js";
import "./paths-BY8fKpqm.js";
import { J as logVerbose, Z as shouldLogVerbose } from "./subsystem-Ck26JAQG.js";
import "./model-selection-BgxWRXtw.js";
import "./github-copilot-token-D8k4aAom.js";
import "./env-BlZ2KU7g.js";
import "./plugins-BZ2misAE.js";
import "./accounts-BoFg1f6e.js";
import "./bindings-D9w3Ca88.js";
import "./accounts-D7QA6mSG.js";
import "./image-ops-DJNnQ3kI.js";
import "./pi-model-discovery-DaNAekda.js";
import "./message-channel-CTEYhii8.js";
import "./pi-embedded-helpers-Bmd0tMWs.js";
import "./config-BgH7L6YG.js";
import "./manifest-registry-RfG50-0E.js";
import "./dock-DmGgAZ_Y.js";
import "./chrome-6GffQ3WS.js";
import "./ssrf-B7akL5oa.js";
import "./skills-DaqYWvx6.js";
import "./redact-DuzbwM25.js";
import "./errors-DPA8_FS3.js";
import "./store-Ccw12g39.js";
import "./sessions-Di5XvcEo.js";
import "./accounts-Ct8GydwU.js";
import "./paths-DbuO2gD6.js";
import "./tool-images-lkYNg3B1.js";
import "./thinking-ZaPrKXBc.js";
import "./image-C2bqxIYC.js";
import "./gemini-auth-DA54LMWP.js";
import "./fetch-guard-CdW8okJa.js";
import "./local-roots-CjfMwc6a.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-Bq-Giyvz.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
