// .ignored_openclaw/dist/extensionAPI.js
import { b as ensureAgentWorkspace, r as resolveAgentDir, s as resolveAgentWorkspaceDir } from "./agent-scope-DKbLFdDh.js";
import "./paths-BY8fKpqm.js";
import "./subsystem-Ck26JAQG.js";
import { h as resolveThinkingDefault, it as DEFAULT_PROVIDER, rt as DEFAULT_MODEL } from "./model-selection-BgxWRXtw.js";
import { a as resolveAgentIdentity } from "./reply-prefix-COPo8CXL.js";
import "./github-copilot-token-D8k4aAom.js";
import "./env-BlZ2KU7g.js";
import "./tokens-BrrydgNN.js";
import { A as resolveAgentTimeoutMs, t as runEmbeddedPiAgent } from "./pi-embedded-BfJSmJ1k.js";
import "./plugins-BZ2misAE.js";
import "./accounts-BoFg1f6e.js";
import "./bindings-D9w3Ca88.js";
import "./send-BH9aKRhH.js";
import "./send-DwdhCvgE.js";
import "./deliver-BMN8TKH9.js";
import "./diagnostic-93YiZ_J1.js";
import "./diagnostic-session-state-Bu3E4Enn.js";
import "./accounts-D7QA6mSG.js";
import "./send-PqUHKJ9p.js";
import "./image-ops-DJNnQ3kI.js";
import "./pi-model-discovery-DaNAekda.js";
import "./message-channel-CTEYhii8.js";
import "./pi-embedded-helpers-Bmd0tMWs.js";
import "./config-BgH7L6YG.js";
import "./manifest-registry-RfG50-0E.js";
import "./dock-DmGgAZ_Y.js";
import "./chrome-6GffQ3WS.js";
import "./ssrf-B7akL5oa.js";
import "./skills-DaqYWvx6.js";
import "./redact-DuzbwM25.js";
import "./errors-DPA8_FS3.js";
import "./store-Ccw12g39.js";
import { s as loadSessionStore, u as saveSessionStore } from "./sessions-Di5XvcEo.js";
import "./accounts-Ct8GydwU.js";
import { l as resolveStorePath, r as resolveSessionFilePath } from "./paths-DbuO2gD6.js";
import "./tool-images-lkYNg3B1.js";
import "./thinking-ZaPrKXBc.js";
import "./image-C2bqxIYC.js";
import "./manager-12qQpt0v.js";
import "./gemini-auth-DA54LMWP.js";
import "./fetch-guard-CdW8okJa.js";
import "./query-expansion-CE4VV_64.js";
import "./retry-CPLC-6FL.js";
import "./target-errors-B6TV2d9b.js";
import "./chunk-BZ9d6HuY.js";
import "./markdown-tables-DmWQZlUR.js";
import "./local-roots-CjfMwc6a.js";
import "./ir-CAlWIb90.js";
import "./render-DwEu-aCr.js";
import "./commands-registry-1VhRTmUY.js";
import "./skill-commands-BZP7JikK.js";
import "./runner-Bq-Giyvz.js";
import "./fetch-CnqePVOV.js";
import "./channel-activity-H1IRpJvl.js";
import "./tables-CzNlNeQv.js";
import "./send-16m7HKVA.js";
import "./outbound-attachment-hwZVuJwx.js";
import "./send-DpcTscqe.js";
import "./resolve-route-DmJwuu1n.js";
import "./proxy-d9fklNTC.js";
import "./replies-DEhDNO2H.js";

export { DEFAULT_MODEL, DEFAULT_PROVIDER, ensureAgentWorkspace, loadSessionStore, resolveAgentDir, resolveAgentIdentity, resolveAgentTimeoutMs, resolveAgentWorkspaceDir, resolveSessionFilePath, resolveStorePath, resolveThinkingDefault, runEmbeddedPiAgent, saveSessionStore };
