// .ignored_openclaw/dist/llm-slug-generator.js
import { c as resolveDefaultAgentId, i as resolveAgentEffectiveModelPrimary, r as resolveAgentDir, s as resolveAgentWorkspaceDir } from "./agent-scope-PyLC-ZZu.js";
import "./paths-C9do7WCN.js";
import { t as createSubsystemLogger } from "./subsystem-9nuH7eAI.js";
import "./workspace-CxArQZC-.js";
import { it as DEFAULT_PROVIDER, l as parseModelRef, rt as DEFAULT_MODEL } from "./model-selection-C6wZ4s0z.js";
import "./github-copilot-token-BkwQAVvU.js";
import "./env-BCSW4P-L.js";
import "./boolean-mcn6kL0s.js";
import "./tokens-CGafcE1g.js";
import { t as runEmbeddedPiAgent } from "./pi-embedded-BDhvoWGL.js";
import "./plugins-D93Gl3vX.js";
import "./accounts-C2p3gWX0.js";
import "./bindings-XIxTfg76.js";
import "./send-DNoo-__b.js";
import "./send-CGHWrWIs.js";
import "./deliver-DMxY-UiH.js";
import "./diagnostic-Bfq8INpY.js";
import "./diagnostic-session-state-C0Sxjfox.js";
import "./accounts-DTMKYA1I.js";
import "./send-WED6JSOM.js";
import "./image-ops-D_zTmnMF.js";
import "./pi-model-discovery-C-yOXpma.js";
import "./message-channel-DSGEYImB.js";
import "./pi-embedded-helpers-BWNHLmOk.js";
import "./config-8KUmruTM.js";
import "./manifest-registry-BoNM8D8y.js";
import "./dock-BQZ3dOVS.js";
import "./chrome-B_oaL_VN.js";
import "./ssrf-BnuIFCPj.js";
import "./frontmatter-rZSrGiqd.js";
import "./skills-BmmA4kHX.js";
import "./redact-Dhp4XjZk.js";
import "./errors-oW2IM9f0.js";
import "./store-D1p2WpL_.js";
import "./sessions-BJ6QL7GG.js";
import "./accounts-CcUE2V_t.js";
import "./paths-CLWDvYDE.js";
import "./tool-images-BYjb6VbO.js";
import "./thinking-CJoHneR6.js";
import "./image-DIjIaXgj.js";
import "./reply-prefix-D9MRO3Gt.js";
import "./manager-CxmUPBMF.js";
import "./gemini-auth-iOIz_-mf.js";
import "./fetch-guard-Py2e9YaA.js";
import "./query-expansion-Co0bWU3Q.js";
import "./retry-Drrk8tTZ.js";
import "./target-errors-COtYfWxr.js";
import "./chunk-CP5JrSGh.js";
import "./markdown-tables-Dc5vS6O9.js";
import "./local-roots-BL_0qvcy.js";
import "./ir-B9Vab0HN.js";
import "./render-loap2gRq.js";
import "./commands-registry-C_No2z8N.js";
import "./skill-commands-D9vTBxFO.js";
import "./runner-BS2WkEnz.js";
import "./fetch-B1nZSYJF.js";
import "./channel-activity-KmaKrhP4.js";
import "./tables-Bf1rDqhy.js";
import "./send-z9O0u_9n.js";
import "./outbound-attachment-D7270zuT.js";
import "./send-Jzg2ogVt.js";
import "./resolve-route-nRTJSWCj.js";
import "./proxy-Bee2aKQk.js";
import "./replies-C9sCMljO.js";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";

//#region src/hooks/llm-slug-generator.ts
/**
* LLM-based slug generator for session memory filenames
*/
const log = createSubsystemLogger("llm-slug-generator");
/**
* Generate a short 1-2 word filename slug from session content using LLM
*/
async function generateSlugViaLLM(params) {
	let tempSessionFile = null;
	try {
		const agentId = resolveDefaultAgentId(params.cfg);
		const workspaceDir = resolveAgentWorkspaceDir(params.cfg, agentId);
		const agentDir = resolveAgentDir(params.cfg, agentId);
		const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "openclaw-slug-"));
		tempSessionFile = path.join(tempDir, "session.jsonl");
		const prompt = `Based on this conversation, generate a short 1-2 word filename slug (lowercase, hyphen-separated, no file extension).

Conversation summary:
${params.sessionContent.slice(0, 2e3)}

Reply with ONLY the slug, nothing else. Examples: "vendor-pitch", "api-design", "bug-fix"`;
		const modelRef = resolveAgentEffectiveModelPrimary(params.cfg, agentId);
		const parsed = modelRef ? parseModelRef(modelRef, DEFAULT_PROVIDER) : null;
		const provider = parsed?.provider ?? DEFAULT_PROVIDER;
		const model = parsed?.model ?? DEFAULT_MODEL;
		const result = await runEmbeddedPiAgent({
			sessionId: `slug-generator-${Date.now()}`,
			sessionKey: "temp:slug-generator",
			agentId,
			sessionFile: tempSessionFile,
			workspaceDir,
			agentDir,
			config: params.cfg,
			prompt,
			provider,
			model,
			timeoutMs: 15e3,
			runId: `slug-gen-${Date.now()}`
		});
		if (result.payloads && result.payloads.length > 0) {
			const text = result.payloads[0]?.text;
			if (text) return text.trim().toLowerCase().replace(/[^a-z0-9-]/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "").slice(0, 30) || null;
		}
		return null;
	} catch (err) {
		const message = err instanceof Error ? err.stack ?? err.message : String(err);
		log.error(`Failed to generate slug: ${message}`);
		return null;
	} finally {
		if (tempSessionFile) try {
			await fs.rm(path.dirname(tempSessionFile), {
				recursive: true,
				force: true
			});
		} catch {}
	}
}

//#endregion
export { generateSlugViaLLM };
